File: c++/14852.cpp
#include <iostream>

typedef long long ll;

using namespace std;

ll dp[10000001];

int main() {
  freopen("example.txt", "r", stdin);
  int N;
  cin >> N;

  dp[0] = 1;
  dp[1] = 2;
  dp[2] = 7;

  ll sum = 0;

  for (int i = 3; i <= N; i++) {
    sum = (sum + dp[i - 3] * 2) % 1000000007;
    dp[i] = (dp[i - 1] * 2 + dp[i - 2] * 3 + sum) % 1000000007;
  }

  cout << dp[N];

  return 0;
}

File: js/14852.js
const fs = require('fs');

const N = +fs.readFileSync('example.txt').toString();

const DIVIDER = 1000000007n;
const dp = Array(1000001).fill(0n);
dp[0] = 1n;
dp[1] = 2n;
dp[2] = 7n;

let sum = 0n;

for (let i = 3; i <= N; i++) {
  sum = (sum + 2n * dp[i - 3]) % DIVIDER;
  dp[i] = (dp[i - 1] * 2n + dp[i - 2] * 3n + sum) % DIVIDER;
}

console.log(dp[N].toString());
